src/include/execution/sql/index_iterator.h

```diff
@@ -0,0 +1,108 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+#include "catalog/catalog_defs.h"
+#include "execution/exec/execution_context.h"
+#include "execution/sql/projected_columns_iterator.h"
+#include "storage/storage_defs.h"
+
+namespace terrier::execution::sql {
+/**
+ * Allows iteration for indices from TPL.
+ */
+class IndexIterator {
+ public:
+  /**
+   * Constructor
+   * @param table_oid oid of the table
+   * @param index_oid oid of the index to iterate over.
+   * @param exec_ctx execution containing of this query
+   * @param col_oids oids of the table columns
+   * @param num_oids number of oids
+   */
+  explicit IndexIterator(uint32_t table_oid, uint32_t index_oid, exec::ExecutionContext *exec_ctx, u32 *col_oids,
+                         u32 num_oids);
+
+  /**
+   * Initialize the projected row and begin scanning.
+   */
+  void Init();
+
+  /**
+   * Frees allocated resources.
+   */
+  ~IndexIterator();
+
+  /**
+   * Wrapper around the index's ScanKey
+   */
+  void ScanKey() {
+    // Scan the index
+    tuples_.clear();
+    curr_index_ = 0;
+    index_->ScanKey(*exec_ctx_->GetTxn(), *index_pr_, &tuples_);
+  }
+
+  /**
+   * Advances the iterator. Return true if successful
+   * @return whether the iterator was advanced or not.
+   */
+  bool Advance() {
+    if (curr_index_ < tuples_.size()) {
+      table_->Select(exec_ctx_->GetTxn(), tuples_[curr_index_], table_pr_);
+      ++curr_index_;
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Get a pointer to the value in the column at index @em col_idx
+   * @tparam T The desired data type stored in the vector projection
+   * @tparam Nullable Whether the column is NULLable
+   * @param col_idx The index of the column to read from
+   * @param[out] null null Whether the given column is null
+   * @return The typed value at the current iterator position in the column
+   */
+  template <typename T, bool Nullable>
+  const T *Get(u16 col_idx, bool *null) const {
+    // NOLINTNEXTLINE: bugprone-suspicious-semicolon: seems like a false positive because of constexpr
+    if constexpr (Nullable) {
+      TERRIER_ASSERT(null != nullptr, "Missing output variable for NULL indicator");
+      *null = table_pr_->IsNull(col_idx);
+    }
+    return reinterpret_cast<T *>(table_pr_->AccessWithNullCheck(col_idx));
```

Comment:
Is there an `AccessWithoutNullCheck` that should be used if `Nullable` is false?

Reply:
There is no `AccessWithoutNullCheck` yet. It can be added we preset the null bitmap to 1 for the right columns, or if we modify the `storage/index` functions to also use `AccessWithoutNullCheck` when possible.
I'd say that's another PR.